EXT#chn=size, ADVAL 7,8,9

jgharston opened the pull request:
EXT#= hands over to I/O interface layer, added dummy setext() to fill in later.
ADVAL 7,8,9 implemented,

jgharston committed:
EXT#c=n, ADVAL 7,8,9

06-Jan-2025
EXT#= hands over to I/O interface layer, added dummy setext() to fill in later.
ADVAL 7,8,9 implemented,

jgharston committed:
EXT#c=n, ADVAL 7,8,9

EXT#= hands over to I/O interface layer, added dummy setext() to fill in later.
ADVAL 7,8,9 implemented,

jgharston committed:
EXT#c=n, ADVAL 7,8,9

EXT#= hands over to I/O interface layer, added dummy setext() to fill in later.
ADVAL 7,8,9 implemented,

rtrussell commented:
To ensure that the proposed changes don't alter existing functionality, shouldn't `setext()` (in both **bbcmos.c** and **bbccon.c**) result in the '_Sorry, not implemented_' error, rather than failing silently?

jgharston committed:
Corrected ADVAL(9)

MOUSE and ADVAL(9) return button state in different order.

jgharston committed:
Initial stab at EXT#chn=num

rtrussell commented:
Needless to say, had an acceptable implementation of `setext` been possible in SDL 2 I would have incorporated it, but that functionality simply isn't supported.  Being able to **extend** a file but not to **truncate** it isn't an acceptable substitute, in my opinion.

jgharston committed:
EXT#chn=size works for Windows

API now matches RISC-V (and similar direct OS_ARGS call) branch.
TODO: Implement for other platforms.

jgharston committed:
EXT#chn=size works for Windows

API now matches RISC-V (and similar direct OS_ARGS call) branch.
Can now continue pulling RISC-V branch closer to master.
TODO: Implement for other platforms.

jgharston committed:
Added __riscv__ tests.

Migrating v1.40 master into RISC-V version. This requires the addition of a few #ifdef __riscv__ directives.

jgharston commented:
I've been trying to catch up on stuff this week.

I've been concentrating on updating the RISC-V branch bringing it up to match v1.40a before digging into deeper code. That has required the main SDL master needing some #ifdef __riscv__ references.

There's one bit though where's I'm uncertain about using "__riscv__", in bbeval.c:

case TMODE:
#ifdef __riscv__
  v.i.n = getmode() ;
#else
  v.i.n = modeno ;
#endif
  v.i.t = 0 ;
  return v ;

as this isn't actually a __riscv__ related thing, but specifying an action for where the VDU driver is not the SDL or Console system incorporated in the BBCBASIC code. So it's a sort-of "use host's VDU driver, not the SDL or Console drivers".

The clearest solution might be to just move all the code within CASE TMODE into each of the actual VDU driver modules, and then the code in bbeval.c simply becomes:
case TMODE:
  v.i.n = getmode() ;
  v.i.t = 0 ;
  return v ;


The filename structure of the RISC-V fork follows the master code:
  bbasmb_riscv.c in the bbasmb_*.c box (1) cluster, and bbdata_riscv.c in the bbdata_*.c box (2) cluster.

"riscv_wrapper.c" is the name Dave Hoglet chose for the I/O functions. Its closest fit is in the "Console" cluster, but it feels out of place in the naming structure. I'm inclined to recommed to Dave that it should be something like "bbriscv.c" to match the naming structure of others, bbpico.c, bbccos.c, etc. See https://mdfs.net/Software/BBCBasic/RISCV/files.png


I had a third thing I was thinking about, but can't remember it just now.

rtrussell commented:
> `case TMODE: v.i.n = getmode() ; v.i.t = 0 ; return v ;`

I don't think that's a contentious proposal, but what are the other places where you felt it necessary to do an `#ifdef riscv`?

(None of my previous comments has received a reply or acknowledgement, so am I talking to myself again?)

rtrussell commented:
The correct pre-processor test to identify a RISC-V target is `#ifdef __riscv` **_not_** `#ifdef riscv` or `#ifdef __riscv__`.  See https://github.com/embecosm/riscv-toolchain-conventions-vendor-relocs

rtrussell commented:
Unfortunately there is a major complication (and incompatibility) in the proposal to implement ADVAL 7.8,9, which is that modern Game Controllers support many more than four axes and it's vitally important that ADVAL can access them.  So, in general, ADVAL(n), where n is positive, must return axis n-1, if supported by the Game Controller.

rtrussell commented:
I have manually incorporated this pull request as follows:

`EXT#chn=size` : Moved from generic interpreter into customisable module.
`lvar=MODE`: Moved from generic interpreter into customisable module.
`ADVAL 7,8,9`: Sorry, couldn't action this request for the reason previously given (Game Controller with many axes).

RISC-V changes are outside the scope of this Pull Request, please open a new one if this remains an issue.
